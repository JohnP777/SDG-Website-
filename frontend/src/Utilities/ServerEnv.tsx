export const getUrl = () => {
    if (process.env.REACT_APP_BACKEND_URL) {
        return process.env.REACT_APP_BACKEND_URL;
    }
    return 'http://127.0.0.1:8000/'
}