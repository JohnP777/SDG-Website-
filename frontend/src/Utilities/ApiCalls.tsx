import { getUrl } from "./ServerEnv";

export const apiCallPost = async (path: string, body: any, isAuthed: boolean ) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'POST',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallGet = async (path: string, isAuthed: boolean, headersExtra?: any) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'GET',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : '',
        ...headersExtra
      },
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallPatch = async (path: string, body: object, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'PATCH',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallDelete = async (path: string, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'DELETE',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
    })
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error)
  }
};

export const apiCallPut = async (path: string, body: any, isAuthed: boolean) => {
  try {
    const response = await fetch(getUrl() + path, {
      method: 'PUT',
      headers: {
        'Content-type': 'application/json',
        'Authorization': isAuthed ? `Token ${localStorage.getItem('token')}` : ''
      },
      body: JSON.stringify(body)
    });
    const data = await response.json();
    return {
      statusCode: response.status,
      ...data
    };
  } catch (error: any) {
    console.log(error);
  }
};
