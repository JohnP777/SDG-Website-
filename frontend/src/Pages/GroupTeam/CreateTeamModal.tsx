import { Box, Typography, InputBase, Paper, Chip } from "@mui/material";
import PageButton, { PageButtonColour } from "../../Components/PageButton";
import PageModal from "../../Components/PageModal";
import React, { useState } from "react";
import { apiCallGet, apiCallPost } from "../../Utilities/ApiCalls";

const CreateTeamModal = () => {
  const [teamName, setTeamName] = useState("");
  const [searchTerm, setSearchTerm] = useState("");
  const [filteredUsers, setFilteredUsers] = useState<string[]>([]);
  const [selectedUsers, setSelectedUsers] = useState<string[]>([]);
  const [message, setMessage] = useState(<></>);
  const [users, setUsers] = useState<string[]>([]);
  const [showFiltered, setShowFiltered] = useState(false);

  // Retrieve initial users
  React.useEffect(() => {
    const fetchData = async () => {
      const data = await apiCallGet('api/teams/users/', true);
      if (data.statusCode === 200) {
        setUsers(data.usernames);
      }
    }
    fetchData();
  }, []);

  // If the searchTerm has already been populated before the users are set
  // asynchronously, then filter.
  //
  // Otherwise filter whenever searchTerm changes.
  React.useEffect(() => {
    if (searchTerm) {
      const filtered = users.filter((user) =>
        user.toLowerCase().includes(searchTerm.toLowerCase())
      );
      setFilteredUsers(filtered);
    } else {
      setFilteredUsers([]);
    }
  }, [users, searchTerm]);

  // Search handler for the search by users feature
  const handleSearchChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;
    setSearchTerm(value);
    setMessage(<></>);
  };

  // Adding/removing user as chips from the list of invited users
  const handleAddUser = (user: string) => {
    if (!selectedUsers.includes(user)) {
      setSelectedUsers([...selectedUsers, user]);
    }
    setSearchTerm("");
    setFilteredUsers([]);
  };

  const handleRemoveUser = (user: string) => {
    setSelectedUsers(selectedUsers.filter((u) => u !== user));
  };

  const handleInputBlur = () => {
    // Add a timeout so that the filteredList is actually clickable
    // Otherwise the target will disappear immediately on blur and nothing
    // will happen
    setTimeout(() => {
      setShowFiltered(false);
    }, 100)
  }

  const handleInputFocus = () => {
    setShowFiltered(true);
  }

  // Send team submission data to backend
  const submitTeam = async () => {
    let data = await apiCallPost('api/teams/create/', {
      'name': teamName,
      'invite_usernames': selectedUsers
    }, true);
    if (data.statusCode === 201) {
      setSelectedUsers([]);
      window.location.reload();
    } else {
      setMessage(
        <Typography textAlign='center' color='error' sx={{wordBreak: 'break-word'}}>
          Error creating team.
        </Typography>
      )
    }
  };

  return (
    <PageModal buttonColour={PageButtonColour.Blue} buttonText="Create a Team">
      <Box
        sx={{
          display: "flex",
          flexDirection: "column",
          justifyContent: "center",
          rowGap: "15px",
        }}
      >
        <Typography variant="h6">Create an SDG Team</Typography>
        <Typography variant="body1" sx={{ paddingTop: 1 }}>
          Team Name
        </Typography>
        <Paper
          component="form"
          sx={{
            p: '2px 4px',
            display: 'flex',
            alignItems: 'center',
            maxWidth: '400px',
            flexGrow: 1,
            position: 'relative',
            '@media (max-width: 805px)': {
              maxWidth: '757px',
              width: 'auto'
            },
          }}
        >
          <InputBase
            sx={{ ml: 1, flex: 1 }}
            placeholder="e.g. Team Green, SDG Project"
            value={teamName}
            onChange={(e) => setTeamName(e.target.value)}
          />
        </Paper>
        <Typography variant="body1" sx={{ paddingTop: 1 }}>
          Who should be in this team?
        </Typography>
        <Paper
          component="form"
          sx={{
            p: '2px 4px',
            display: 'flex',
            alignItems: 'center',
            maxWidth: '400px',
            flexGrow: 1,
            position: 'relative',
            '@media (max-width: 805px)': {
              maxWidth: '757px',
              width: 'auto'
            },
          }}
        >
          <InputBase
            sx={{ ml: 1, flex: 1 }}
            placeholder="Type the username of a friend"
            value={searchTerm}
            onChange={handleSearchChange}
            onBlur={handleInputBlur}
            onFocus={handleInputFocus}
          />
          {/* Dropdown menu with suggested users based on search query */}
          {showFiltered && filteredUsers.length > 0 && (
          <Paper
            sx={{
              position: "absolute",
              top: "100%",
              left: 0,
              width: "100%",
              backgroundColor: "#fff",
              zIndex: 10,
              mt: "2px",
              boxShadow: 3,
              borderRadius: "4px",
              maxHeight: 150,
              overflowY: "auto",
            }}
          >
            {filteredUsers.map((user) => (
              <Typography
                key={user}
                sx={{
                  padding: 1,
                  cursor: "pointer",
                  "&:hover": { backgroundColor: "#f0f0f0" },
                }}
                onClick={() => handleAddUser(user)}
              >
                {user}
              </Typography>
            ))}
          </Paper>
        )}
        </Paper>
        {/* Chips that represent selected users */}
        {selectedUsers.length > 0 && (
          <Box sx={{ display: "flex", flexWrap: "wrap", gap: 1 }}>
            {selectedUsers.map((user) => (
              <Chip
                key={user}
                label={user}
                onDelete={() => handleRemoveUser(user)}
                sx={{ backgroundColor: "#d2e8fc" }}
              />
            ))}
          </Box>
        )}
        {message}
        <Box sx={{ display: "flex", justifyContent: "space-around" }}>
          <PageButton colour={PageButtonColour.Blue} onClick={submitTeam}>
            Create Team
          </PageButton>
        </Box>
      </Box>
    </PageModal>
  );
};

export default CreateTeamModal;
