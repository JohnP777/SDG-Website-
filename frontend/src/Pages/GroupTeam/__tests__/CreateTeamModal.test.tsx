import { render, screen, fireEvent, waitFor, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import CreateTeamModal from "../CreateTeamModal";
import * as ApiCalls from "../../../Utilities/ApiCalls";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";

// Mock the API calls
jest.mock("../../../Utilities/ApiCalls", () => ({
  apiCallGet: jest.fn(),
  apiCallPost: jest.fn(),
}));

// Mock window.location.reload
const mockReload = jest.fn();
Object.defineProperty(window, 'location', {
  value: {
    reload: mockReload,
  },
  writable: true
});

// Helper function to render the component with router
const renderWithRouter = (component: React.ReactNode) => {
  return render(
    <MemoryRouter>
      {component}
    </MemoryRouter>
  );
};

describe("CreateTeamModal", () => {
  // Reset mocks before each test
  beforeEach(() => {
    jest.clearAllMocks();
    (ApiCalls.apiCallGet as jest.Mock).mockResolvedValue({
      statusCode: 200,
      usernames: ['user1', 'user2', 'user3'],
    });
  });

  // Snapshot test
  it("should match snapshot", async () => {
    const { asFragment } = renderWithRouter(<CreateTeamModal />);
    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });

  // Test modal opens when button is clicked
  it("should open modal when button is clicked", async () => {
    renderWithRouter(<CreateTeamModal />);
    
    // Find and click the "Create a Team" button
    const createButton = screen.getByText("Create a Team");
    fireEvent.click(createButton);
  
    // Check if modal content is visible
    await waitFor(() => {
      expect(screen.getByText("Create an SDG Team")).toBeInTheDocument();
    })
  });

  // Test team name input
  it("should update team name when input changes", async () => {
    renderWithRouter(<CreateTeamModal />);
    
    // Open the modal
    const createButton = screen.getByText("Create a Team");
    fireEvent.click(createButton);
    
    // Find the team name input and change its value
    const teamNameInput = screen.getByPlaceholderText("e.g. Team Green, SDG Project");
    fireEvent.change(teamNameInput, { target: { value: "Test Team" } });
    
    // Check if the input value has been updated
    await waitFor(() => {
      expect(teamNameInput).toHaveValue("Test Team");
    })
  });

  // Test user search functionality
  it("should filter users when searching", async () => {
    renderWithRouter(<CreateTeamModal />);
    
    // Open the modal
    const createButton = screen.getByText("Create a Team");
    fireEvent.click(createButton);
    
    // Find the user search input and type in it
    const searchInput = screen.getByPlaceholderText("Type the username of a friend");
    userEvent.type(searchInput, "user");

    // Check if input value has been updated
    expect(searchInput).toHaveValue("user");
    
    // Wait for the filtered users to appear
    await waitFor(() => {
      expect(screen.getByText("user1")).toBeInTheDocument();
      expect(screen.getByText("user2")).toBeInTheDocument();
      expect(screen.getByText("user3")).toBeInTheDocument();
    });
  });

  // Test user search on blur
  it("should remove user suggestions on blur", async () => {
    renderWithRouter(<CreateTeamModal />);
    
    // Open the modal
    const createButton = screen.getByText("Create a Team");
    fireEvent.click(createButton);
    
    // Find the user search input and type in it
    const searchInput = screen.getByPlaceholderText("Type the username of a friend");
    userEvent.type(searchInput, "user");
    
    // Wait for the filtered users to appear
    await waitFor(() => {
      expect(screen.getByText("user1")).toBeInTheDocument();
    });

    fireEvent.blur(searchInput);
    // Wait for the filtered users to appear
    await waitFor(() => {
      expect(screen.queryByText("user1")).not.toBeInTheDocument();
    });
  });

  // Test adding a user
  it("should add a user when clicked from search results", async () => {
    renderWithRouter(<CreateTeamModal />);
    
    // Open the modal
    const createButton = screen.getByText("Create a Team");
    fireEvent.click(createButton);
    
    // Search for users
    const searchInput = screen.getByPlaceholderText("Type the username of a friend");
    userEvent.type(searchInput, "user");
    
    // Wait for search results and click on a user
    await waitFor(() => {
      const user1 = screen.getByText("user1");
      fireEvent.click(user1);
    });

    await waitFor(() => {
      // Check if the user has been added (as a chip)
      expect(screen.getByText("user1")).toBeInTheDocument();
      
      // Check if the search input has been cleared
      expect(searchInput).toHaveValue("");
    })
  });

  // Test removing a user
  it("should remove a user when delete button on chip is clicked", async () => {
    renderWithRouter(<CreateTeamModal />);
    
    // Open the modal
    const createButton = screen.getByText("Create a Team");
    fireEvent.click(createButton);
    
    // Search for users and add one
    const searchInput = screen.getByPlaceholderText("Type the username of a friend");
    userEvent.type(searchInput, "user");
    
    await waitFor(() => {
      const user1 = screen.getByText("user1");
      fireEvent.click(user1);
    });
    
    // Find the chip with the user and click its delete button
    const chip = screen.getByText("user1").closest(".MuiChip-root");
    const deleteButton = chip?.querySelector(".MuiChip-deleteIcon");
    if (deleteButton) {
      fireEvent.click(deleteButton);
    }
    
    // Check if the user has been removed
    await waitFor(() => {
      expect(screen.queryByText("user1")).not.toBeInTheDocument();
    });
  });
});