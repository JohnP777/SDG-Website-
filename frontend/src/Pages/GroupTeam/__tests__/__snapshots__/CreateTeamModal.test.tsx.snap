// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTeamModal should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1317kya-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Create a Team
    </button>
  </div>
</DocumentFragment>
`;
